File: VulkanGameEngine/RenderMesh2DComponent.cpp
#include "RenderMesh2DComponent.h"
#include "MemoryManager.h"

RenderMesh2DComponent::RenderMesh2DComponent() : GameObjectComponent()
{
}

RenderMesh2DComponent::RenderMesh2DComponent(String name, uint32 meshBufferIndex) : GameObjectComponent(name, ComponentTypeEnum::kRenderMesh2DComponent)
{
	std::vector<Vertex2D> SpriteVertexList =
	{
	  { {0.0f, 0.5f},  {0.0f, 0.0f}, {1.0f, 0.0f, 0.0f, 1.0f} },
	  { {0.5f, 0.5f},  {1.0f, 0.0f}, {0.0f, 1.0f, 0.0f, 1.0f} },
	  { {0.5f, 0.0f},  {1.0f, 1.0f}, {0.0f, 0.0f, 1.0f, 1.0f} },
	  { {0.0f, 0.0f},  {0.0f, 1.0f}, {1.0f, 1.0f, 0.0f, 1.0f} }
	};
	std::vector<uint32> SpriteIndexList =
	{
	  0, 1, 3,
	  1, 2, 3
	};
	mesh = std::make_shared<Mesh2D>(Mesh2D(SpriteVertexList, SpriteIndexList, meshBufferIndex));
}

RenderMesh2DComponent::~RenderMesh2DComponent()
{
}

std::shared_ptr<RenderMesh2DComponent> RenderMesh2DComponent::CreateRenderMesh2DComponent(String name, uint32 meshBufferIndex)
{
	std::shared_ptr<RenderMesh2DComponent> gameObject = MemoryManager::AllocateRenderMesh2DComponent();
	new (gameObject.get()) RenderMesh2DComponent(name, meshBufferIndex);
	return gameObject;
}

void RenderMesh2DComponent::Update(float deltaTime)
{
	mesh->Update(deltaTime);
}

void RenderMesh2DComponent::BufferUpdate(VkCommandBuffer& commandBuffer, float deltaTime)
{
	mesh->BufferUpdate(commandBuffer, deltaTime);
	mesh->Update(deltaTime);
}

void RenderMesh2DComponent::Draw(VkCommandBuffer& commandBuffer, VkPipeline& pipeline, VkPipelineLayout& shaderPipelineLayout, VkDescriptorSet& descriptorSet, SceneDataBuffer& sceneProperties)
{
	mesh->Draw(commandBuffer, pipeline, shaderPipelineLayout, descriptorSet, sceneProperties);
}

void RenderMesh2DComponent::Destroy()
{
	mesh->Destroy();
}

std::shared_ptr<GameObjectComponent> RenderMesh2DComponent::Clone() const
{
	return std::make_shared<RenderMesh2DComponent>(*this);
}

size_t RenderMesh2DComponent::GetMemorySize() const
{
	return sizeof(RenderMesh2DComponent);
}

File: VulkanGameEngineLevelEditor/Components/RenderMesh2DComponent.cs
using Coral.Managed.Interop;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using VulkanGameEngineGameObjectScripts.Input;
using VulkanGameEngineGameObjectScripts;
using VulkanGameEngineLevelEditor.GameEngineAPI;

namespace VulkanGameEngineLevelEditor.Components
{
    public unsafe class RenderMesh2DComponent : GameObjectComponent
    {
        public RenderMesh2DComponent() : base()
        {
            Name = "RenderMesh2DComponent";
        }

        public RenderMesh2DComponent(GameObjectComponent cppComponentPtr, GameObject cppGameObjectPtr, GameObject csParentGameObject) :
            base(cppComponentPtr, cppGameObjectPtr, csParentGameObject, ComponentTypeEnum.kRenderMesh2DComponent)
        {
            Name = "RenderMesh2DComponent";
        }

        public RenderMesh2DComponent(GameObjectComponent cppComponentPtr, GameObject cppGameObjectPtr, GameObject csParentGameObject, string name) :
            base(cppComponentPtr, cppGameObjectPtr, csParentGameObject, name, ComponentTypeEnum.kRenderMesh2DComponent)
        {

        }

        public override void Input(KeyBoardKeys key, float deltaTime)
        {
        }

        public override void Update(VkCommandBuffer commandBuffer, float deltaTime)
        {

        }

        public override void Draw(VkCommandBuffer commandBuffer, VkPipeline pipeline, VkPipelineLayout pipelineLayout, VkDescriptorSet descriptorSet, SceneDataBuffer sceneProperties)
        {

        }

        public override void Destroy()
        {

        }

        public override int GetMemorySize()
        {
            return (int)sizeof(RenderMesh2DComponent);
        }
    }
}
